Examples/Nodify.Shapes/Canvas/CanvasViewModel.cs

```diff
@@ -0,0 +1,249 @@
+        public ICommand DeleteSelectionCommand { get; }
+
+        public ICommand UndoCommand { get; }
+        public ICommand RedoCommand { get; }
+
+        public NodifyObservableCollection<UserCursorViewModel> Cursors { get; } = new NodifyObservableCollection<UserCursorViewModel>();
+        public IActionsHistory UndoRedo { get; } = ActionsHistory.Global;
+
+        private readonly Random _rand = new Random();
+
+        public CanvasViewModel()
+        {
+            CanvasToolbar = new CanvasToolbarViewModel(this);
+
+            UndoCommand = new RequeryCommand(UndoRedo.Undo, () => UndoRedo.CanUndo);
+            RedoCommand = new RequeryCommand(UndoRedo.Redo, () => UndoRedo.CanRedo);
+
+            MoveShapesStartedCommand = new DelegateCommand(MoveShapesStartedHandler);
+            MoveShapesCompletedCommand = new DelegateCommand(MoveShapesCompletedHandler);
+
+            SelectShapesStartedCommand = new DelegateCommand(SelectShapesStartedHandler);
+            SelectShapesCompletedCommand = new DelegateCommand(SelectShapesCompletedHandler);
+
+            ResizeShapeStartedCommand = new DelegateCommand(ResizeShapeStartedHandler);
+            ResizeShapeCompletedCommand = new DelegateCommand(ResizeShapeCompletedHandler);
+
+            DeleteSelectionCommand = new DelegateCommand(DeleteSelection, () => !CanvasToolbar.Locked);
+
+            CreateConnectionCommand = new DelegateCommand<(object Source, object Target)>(
+                ((object Source, object Target) pendingConnection) => AddConnection((ConnectorViewModel)pendingConnection.Source, (ConnectorViewModel)pendingConnection.Target),
+                ((object Source, object Target) pendingConnection) => CanConnect((ConnectorViewModel)pendingConnection.Source, (ConnectorViewModel)pendingConnection.Target));
+
+            RemoveConnectionCommand = new DelegateCommand<ConnectionViewModel>(RemoveConnection);
+
+            SelectedShapes.WhenAdded(shape =>
+            {
+                ShapeToolbar.Shape = SelectedShapes.Count == 1 ? shape : null;
+            });
+
+            SelectedShapes.WhenRemoved(shape =>
+            {
+                ShapeToolbar.Shape = SelectedShapes.Count == 1 ? SelectedShapes.Single() : null;
+            });
+
+            SelectedShapes.WhenCleared(shapes => ShapeToolbar.Shape = null);
+
+            FillCanvasWithShapes();
+        }
+
+        private void MoveShapesStartedHandler()
+        {
+            UndoRedo.ExecuteAction(new MoveShapesAction(this));
+            ShapeToolbar.Hide();
+        }
+
+        private void MoveShapesCompletedHandler()
+        {
+            if (UndoRedo.Current is MoveShapesAction movesShapes)
+            {
+                movesShapes.SaveLocations();
+            }
+
+            ShapeToolbar.Show();
+        }
+
+        // TODO: This does not fire for single selection (ItemContainer click), therefore I added IsSelected to ShapeViewModel.
```

Comment:
This is nice for the sample, and very adjustable - for actual practical usage selection themselves being an undo/redo action is likely confusing.

Reply:
Yes, it depends on the business requirements. For example, Figma lets you undo/redo selection. Not a priority right now, but I would like to fix this in the future.